### PR TITLE
Add a function in MooseApp to obtain all checkpoint directories

### LINKED ISSUE #11069: Split MooseApp::getCheckpointFiles()
## Rationale
We want to split `std::list<std::string> MooseApp::getCheckpointFiles()` into two parts, one obtain the checkpoint directories and another for finding all files under the directories. The reason we want to have this split is that we want to use a function like `std::list<std::string> MooseApp::getCheckpointDirectories()` to store an auxiliary file used for recovery. 

## Description

We really do not want to do this but have to. We have some restartable data required for building a material, but the data will not be available during material construction. The data is quite simple so we can just output it into a separate file under recovery folders during construction and can be loaded later in recovery mode. If you are curious, the data is material IDs, related with mesh.
Any comments or better suggestions?

## Impact
New interface function.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #11069.

### CODE REVIEW

**Comment by YaqiWang** [comment]
@rwcarlsen and @permcody maybe you two care about this PR and would like to review it. The failure should be irrelevant with this PR.

File: framework/src/base/MooseApp.C [hunk 1]
```diff
@@ -877,6 +877,13 @@ MooseApp::getCheckpointFiles()
     }
   }
 
+  return checkpoint_dirs;
+}
+
+std::list<std::string>
+MooseApp::getCheckpointFiles() const
```
Comment (rwcarlsen) [inline framework/src/base/MooseApp.C]: With this change of behavior you will need to update all places that called getCheckpointFiles to call getCheckpointDirectories.

Comment (YaqiWang) [reply framework/src/base/MooseApp.C]: getCheckpointDirectories is called inside getCheckpointFiles. Did i miss something?

**Comment by YaqiWang** [comment]
This is ready.

File: framework/src/base/MooseApp.C (hunk 1 above)
Comment (rwcarlsen) [reply framework/src/base/MooseApp.C]: Is this function actually used anywhere?  If not we should just remove it.  Did you need it in rattlesnake - if so you could just implement its two lines in rattlesnake instead.

Comment (YaqiWang) [reply framework/src/base/MooseApp.C]: It is used in framework: `moose/framework/src/actions/SetupRecoverFileBaseAction.C`. Rattlesnake is not using it.

**Comment by rwcarlsen** [comment]
If you look in SetupRecoverFileBaseAction.C  there is code that calls the ``getCheckpointFiles`` function - and this PR changes that behavior.  There may be cases where this causes recovery to not work right.  I think you need to change that call to use ``getCheckpointDirectories`` - and then we shouldn't need ``getCheckpointFiles`` anymore.

**Review by rwcarlsen — APPROVED** [review]
I see how this works now.

**Pull request merged**